src/main/java/org/opensearch/ad/feature/SearchFeatureDao.java

```diff
@@ -190,67 +198,200 @@ public void getHighestCountEntities(AnomalyDetector detector, long startTime, lo
             .includeUpper(false);
 
         BoolQueryBuilder boolQueryBuilder = QueryBuilders.boolQuery().filter(rangeQuery).filter(detector.getFilterQuery());
-        TermsAggregationBuilder termsAgg = AggregationBuilders.terms(AGG_NAME_TERM).size(maxEntitiesForPreview);
-        if (detector.getCategoryField() == null || detector.getCategoryField().isEmpty()) {
-            listener.onResponse(null);
-            return;
-        }
+        AggregationBuilder bucketAggs = null;
 
         if (detector.getCategoryField().size() == 1) {
-            termsAgg.field(detector.getCategoryField().get(0));
+            bucketAggs = AggregationBuilders.terms(AGG_NAME_TOP).size(maxEntitiesForPreview).field(detector.getCategoryField().get(0));
         } else {
-            termsAgg.script(ScriptMaker.makeTermsScript(detector.getCategoryField()));
+            /*
+             * We don't have an efficient solution for terms aggregation on multiple fields.
+             * Terms aggregation does not support collecting terms from multiple fields in the same document.
+             *  We have to work around the limitation by using a script to retrieve terms from multiple fields.
+             *  The workaround disables the global ordinals optimization and thus causes a markedly longer
+             *  slowdown. This is because scripting is tugging on memory and has to iterate through
+             *  all of the documents at least once to create run-time fields.
+             *
+             *  We evaluated composite and terms aggregation using a generated data set with one
+             *  million entities.  Each entity has two documents. Composite aggregation finishes
+             *  around 40 seconds.  Terms aggregation performs differently on different clusters.
+             *  On a 3 data node cluster, terms aggregation does not finish running within 2 hours
+             *  on a 5 primary shard index. On a 15 data node cluster, terms  aggregation needs 217 seconds
+             *  on a 15 primary shard index. On a 30 data node cluster, terms aggregation needs 47 seconds
+             *  on a 30 primary shard index.
+             *
+             * Here we work around the problem using composite aggregation. Composite aggregation cannot
+             * give top entities without collecting all aggregated results. Paginated results are returned
+             * in the natural order of composite keys. This is fine for Preview API. Preview API needs the
+             * top entities to make sure there is enough data for training and showing the results. We
+             * can paginate entities and filter out entities that do not have enough docs (e.g., 256 docs).
+             * As long as we have collected the desired number of entities (e.g., 5 entities), we can stop
```

Comment:
How about add max pagination limit to avoid scanning all entities if no entity has enough docs?

Reply:
how about adding 1 minute expiration time as preview is time-sensitive? If we can finish within time, scanning all entities are acceptable.

Reply:
I'm ok for that

Reply:
please check the upcoming commit